frontend/src/globals/unity-webrequesthandler2.ts

```diff
@@ -0,0 +1,210 @@
+		const resolve = (data) => {
+			if (this.requests[id].abort) {
+				throw new Error('Aborted'); // To be picked up by catch below
+			}
+			this.requests[id].data = data;
+			this.sendOnWebResponse({
+				id,
+				size: data.byteLength,
+			});
+		};
+
+		// Attempt to get the key from the cache. If this is successful it can be
+		// given directly to the viewer, otherwise retrieve it.
+		this.readCache(url, cache).then((result) => {
+			if (result !== undefined) {
+				resolve(result);
+				return Promise.resolve();
+			}
+
+			return fetch(this.getUrl(url)).then((response) => response.arrayBuffer()).then((data) => { // This promise is returned so that any exceptions it produces are caught below
+				this.writeCache(url, data);
+				resolve(data);
+			});
+		}).catch((err) => {
+			if (err.message !== 'Aborted') { // This we don't need to print because in this case the control flow is expected
+				console.log(err);
+			}
+			this.sendOnWebResponse({
+				id,
+				size: -1, // This signals that the request ended early. The reason should be printed in the Js log so there is no need to send it to Unity.
+			});
+		});
+	}
+
+	/** Given a local URI that can also act as a cache key, return a fully
+	 * qualified URI that will return the resource from .io */
+	// eslint-disable-next-line class-methods-use-this
+	getUrl(url) {
+		// In the frontend, we rely on the XHR system to already be set up
+		return `${this.apiHost}/${url}`;
+	}
+
+	/**
+	 * Aborts the request. This will only take effect if sendOnWebResponse has
+	 * not already been called.
+	 */
+	abortRequest(id: number) {
+		this.requests[id].abort = true;
+	}
+
+	// When a byte[] array (or other managed object) is passed to an imported
+	// function, it arrives as an offset into the managed Heap. We can use this
+	// to access the contents of the array directly in the Module's heap
+	// ArrayBuffer, to get or set the contents.
+
+	/**
+	 * Copies the data returned from a get request into the heap at offset.
+	 * Once this is done the local resources should be released by calling
+	 * releaseTransaction.
+	 */
+	getResponseBytes(id: number, offset: number) {
+		try {
+			this.unityInstance.Module.HEAP8.set( // set copies the contents of the TypedArray into the destination ArrayBuffer
+				new Int8Array(this.requests[id].data), // Specify the Type to interpet the ArrayBuffer as when copying
+
+				// This 'shift' converts offset from a signed to unsigned
+				// integer, which is necessary to support the WASM 4GB address
+				// space https://v8.dev/blog/4gb-wasm-memory
+				// eslint-disable-next-line no-bitwise
+				offset >>> 0,
+			);
+		} catch (error) {
+			console.log(error);
+		}
+	}
+
+	/** Converts the data returned from a get request into a string and returns
+	 * it. The plugin will allocate an unmanaged buffer for the string and copy
+	 * it into it.
+	 */
+	getResponseString(id: number): string {
+		const body = new TextDecoder().decode(this.requests[id].data);
```

Comment:
Tiny thing, but we can just return since we don't use the body here.

```suggestion
return new TextDecoder().decode(this.requests[id].data);
```

Reply:
We're sure that there's no situation where we'd go index out of bounds?

Just quickly checked the unity code we also don't catch any error.

If we're confident it's all good, just playing devil's advocate here!

Reply:
Hi @carmenfan, 

Not that I can think of - the member is set in the `createGetRequest` method, which is synchronous, and only deleted by `releaseRequest`, which is also a synchronous call from Unity (i.e. the js side doesn't get any say over what this is or when its updated). Additionally, the control flow is implemented as a co-routine, which only suspends execution, between frames, so there is no way to change the order of operations.

My thinking is more towards the fail fast, fail early approach for this: if we ever went out of bounds here it would not only be an unrecoverable error, but might also call into question whether we've actually been returning the right data for other requests, so probably not something we want to try and recover silently from (and that's if corrupted data hasn't already resulted in a load of failures elsewhere).